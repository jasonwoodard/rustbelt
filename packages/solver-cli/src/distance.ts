import type { Coord } from './types';

/**
 * Compute pairwise Euclidean distances between a list of IDs.
 */
export function pairwiseDistances(
  ids: string[],
  idToCoord: Record<string, Coord>
): number[][] {
  const distances: number[][] = [];

  for (let i = 0; i < ids.length; i++) {
    const coordI = idToCoord[ids[i]];
    if (!coordI) {
      throw new Error(`Missing coordinates for id "${ids[i]}"`);
    }

    distances[i] = [];

    for (let j = 0; j < ids.length; j++) {
      const coordJ = idToCoord[ids[j]];
      if (!coordJ) {
        throw new Error(`Missing coordinates for id "${ids[j]}"`);
      }

      const dx = coordI[0] - coordJ[0];
      const dy = coordI[1] - coordJ[1];
      distances[i][j] = Math.hypot(dx, dy);
    }
  }

  return distances;
}

export function minutesAtMph(distance: number, mph: number): number {
  if (mph <= 0) {
    throw new Error(`Speed must be greater than 0 mph: ${mph}`);
  }
  return (distance / mph) * 60;
}

// Haversine formula to compute great-circle distance between two points on Earth in miles
export function haversineMiles(a: Coord, b: Coord): number {
  const toRad = (deg: number) => (deg * Math.PI) / 180;
  const lat1: number = toRad(a[0]);
  const lon1: number = toRad(a[1]);
  const lat2: number = toRad(b[0]);
  const lon2: number = toRad(b[1]);
  const dLat = lat2 - lat1;
  const dLon = lon2 - lon1;
  const h =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLon / 2) ** 2;
  const R = 3958.8; // Earth radius in miles
  return 2 * R * Math.asin(Math.sqrt(h));
}

/**
 * Build a symmetric distance matrix in miles for the provided coordinates.
 * Only the upper triangle (j > i) is computed and mirrored to the lower triangle.
 */
export function buildMatrix(coords: Coord[]): number[][] {
  const n = coords.length;
  const matrix: number[][] = Array.from({ length: n }, () => Array(n).fill(0));

  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      const dist = haversineMiles(coords[i], coords[j]);
      matrix[i][j] = dist;
      matrix[j][i] = dist; // mirror to lower triangle
    }
  }

  return matrix;
}

/**
 * Compute drive minutes for a given distance and speed, inflated by a
 * robustness factor to make the schedule more conservative.
 */
export function driveMinutes(
  distance: number,
  mph: number,
  robustnessFactor = 1,
): number {
  return minutesAtMph(distance, mph) * robustnessFactor;
}

