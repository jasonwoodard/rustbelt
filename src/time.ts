export function hhmmToMin(time: string): number {
  const [hh, mm] = time.split(':').map(Number);
  return hh * 60 + mm;
}

export function minToHhmm(min: number): string {
  const h = Math.floor(min / 60);
  const m = Math.round(min % 60);
  return `${String(h).padStart(2, '0')}:${String(m).padStart(2, '0')}`;
}
