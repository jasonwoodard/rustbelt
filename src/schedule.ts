import { minutesAtMph, haversineMiles } from './distance';
import type {
  Anchor,
  Store,
  ID,
  Coord,
  StopPlan,
} from './types';

export interface ScheduleCtx {
  start: Anchor;
  end: Anchor;
  window: { start: string; end: string };
  mph: number;
  defaultDwellMin: number;
  stores: Record<ID, Store>;
  mustVisitIds?: ID[];
}

export interface TimelineResult {
  stops: StopPlan[];
  totalDriveMin: number;
  totalDwellMin: number;
  hotelETAmin: number;
}

function hhmmToMin(time: string): number {
  const [hh, mm] = time.split(':').map(Number);
  return hh * 60 + mm;
}

function minToHhmm(min: number): string {
  const h = Math.floor(min / 60);
  const m = Math.round(min % 60);
  return `${String(h).padStart(2, '0')}:${String(m).padStart(2, '0')}`;
}

function legMetrics(
  from: Coord,
  to: Coord,
  mph: number,
): { dist: number; driveMin: number } {
  const dist = haversineMiles(from, to);
  return { dist, driveMin: minutesAtMph(dist, mph) };
}

export function computeTimeline(order: ID[], ctx: ScheduleCtx): TimelineResult {
  const startMin = hhmmToMin(ctx.window.start);
  const stops: StopPlan[] = [];

  let totalDriveMin = 0;
  let totalDwellMin = 0;
  let currentTime = startMin;
  let currentId: ID = ctx.start.id;
  let currentCoord: Coord = ctx.start.coord;

  // start stop
  stops.push({
    id: ctx.start.id,
    type: 'start',
    arrive: minToHhmm(currentTime),
    depart: minToHhmm(currentTime),
  });

  for (const id of order) {
    const store = ctx.stores[id];
    if (!store) {
      throw new Error(`Unknown store id: ${id}`);
    }

    const { dist, driveMin } = legMetrics(
      currentCoord,
      store.coord,
      ctx.mph,
    );
    currentTime += driveMin;
    totalDriveMin += driveMin;

    const dwell = store.dwellMin ?? ctx.defaultDwellMin;
    const arriveMin = currentTime;
    currentTime += dwell;
    totalDwellMin += dwell;

    stops.push({
      id: store.id,
      type: 'store',
      arrive: minToHhmm(arriveMin),
      depart: minToHhmm(currentTime),
      dwellMin: dwell,
      legIn: {
        fromId: currentId,
        toId: store.id,
        driveMin,
        distanceMi: dist,
      },
    });

    currentId = store.id;
    currentCoord = store.coord;
  }

  // leg to end
  const { dist, driveMin } = legMetrics(
    currentCoord,
    ctx.end.coord,
    ctx.mph,
  );
  currentTime += driveMin;
  totalDriveMin += driveMin;

  stops.push({
    id: ctx.end.id,
    type: 'end',
    arrive: minToHhmm(currentTime),
    depart: minToHhmm(currentTime),
    legIn: {
      fromId: currentId,
      toId: ctx.end.id,
      driveMin,
      distanceMi: dist,
    },
  });

  return { stops, totalDriveMin, totalDwellMin, hotelETAmin: currentTime };
}

export function isFeasible(order: ID[], ctx: ScheduleCtx): boolean {
  if (ctx.mustVisitIds) {
    for (const id of ctx.mustVisitIds) {
      if (!order.includes(id)) {
        return false;
      }
    }
  }

  const { hotelETAmin } = computeTimeline(order, ctx);
  const endMin = hhmmToMin(ctx.window.end);
  return hotelETAmin <= endMin;
}

export function slackMin(order: ID[], ctx: ScheduleCtx): number {
  const { hotelETAmin } = computeTimeline(order, ctx);
  const endMin = hhmmToMin(ctx.window.end);
  return Math.max(0, endMin - hotelETAmin);
}

