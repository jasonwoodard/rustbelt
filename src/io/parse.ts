import { OpenLocationCode } from 'open-location-code';
import type {
  Anchor,
  Coord,
  DayConfig,
  Store,
  TripConfig,
  LockSpec,
  Weekday,
  StoreOpenHours,
} from '../types';
import type { TripInput } from '../types';
import { haversineMiles } from '../distance';
import { hhmmToMin } from '../time';

function ensureValidCoord(lat: number, lon: number): Coord {
  if (
    Number.isNaN(lat) ||
    Number.isNaN(lon) ||
    lat < -90 ||
    lat > 90 ||
    lon < -180 ||
    lon > 180
  ) {
    throw new Error(`Invalid coordinates: ${lat},${lon}`);
  }
  return [lat, lon];
}

/**
 * Parse a location string into `[lat, lon]` coordinates.
 * Supports `lat,lon`, Plus Codes, and Google Maps URLs containing `@lat,lon`.
 */
export function parseLocation(input: string): Coord {
  if (typeof input !== 'string') {
    throw new Error('Location must be a string');
  }
  const str = input.trim();

  // Direct "lat,lon" string
  const latLon = str.match(/^(-?\d+(?:\.\d+)?)[,\s]+(-?\d+(?:\.\d+)?)(?:.*)$/);
  if (latLon) {
    const lat = parseFloat(latLon[1]);
    const lon = parseFloat(latLon[2]);
    return ensureValidCoord(lat, lon);
  }

  // Google Maps URL with @lat,lon
  const urlMatch = str.match(/@(-?\d+(?:\.\d+)?),(-?\d+(?:\.\d+)?)/);
  if (urlMatch) {
    const lat = parseFloat(urlMatch[1]);
    const lon = parseFloat(urlMatch[2]);
    return ensureValidCoord(lat, lon);
  }

  // Plus Code
  try {
    const decoded = OpenLocationCode.decode(str);
    return ensureValidCoord(
      decoded.latitudeCenter,
      decoded.longitudeCenter,
    );
  } catch {
    /* noop */
  }

  throw new Error(
    `Unable to parse location: "${input}". Provide lat/lon, a full Plus Code, or a Maps URL with @lat,lon.`,
  );
}

type PlainObj = Record<string, unknown>;

const MISSING_COORDS_MSG = 'missing coordinates';

function parseCoord(obj: PlainObj): Coord {
  if (typeof obj.lat === 'number' && typeof obj.lon === 'number') {
    return ensureValidCoord(obj.lat, obj.lon);
  }
  if (typeof obj.location === 'string') {
    return parseLocation(obj.location);
  }
  throw new Error(MISSING_COORDS_MSG);
}

function parseAnchor(obj: PlainObj): Anchor {
  if (!obj || typeof obj.id !== 'string') {
    throw new Error('Anchor must have an id');
  }
  const id = obj.id;
  const name = typeof obj.name === 'string' ? obj.name : id;
  let coord: Coord;
  try {
    coord = parseCoord(obj);
  } catch (err) {
    if ((err as Error).message === MISSING_COORDS_MSG) {
      throw new Error(`Anchor ${id} missing coordinates`);
    }
    throw err;
  }
  return { id, name, coord };
}

function parseStore(obj: PlainObj): Store {
  if (!obj || typeof obj.id !== 'string') {
    throw new Error('Store must have an id');
  }
  const id = obj.id;
  const name = typeof obj.name === 'string' ? obj.name : id;
  let coord: Coord;
  try {
    coord = parseCoord(obj);
  } catch (err) {
    if ((err as Error).message === MISSING_COORDS_MSG) {
      throw new Error(`Store ${id} missing coordinates`);
    }
    throw err;
  }

  const store: Store = { id, name, coord };

  if (typeof obj.address === 'string') {
    store.address = obj.address;
  }

  if (obj.dwellMin !== undefined) {
    const dwell = Number(obj.dwellMin);
    if (!Number.isFinite(dwell) || dwell < 0) {
      throw new Error(`Invalid dwellMin for store ${id}`);
    }
    store.dwellMin = dwell;
  }

  if (obj.score !== undefined) {
    const score = Number(obj.score);
    if (Number.isFinite(score)) {
      store.score = score;
    }
  }

  if (obj.tags) {
    if (Array.isArray(obj.tags)) {
      store.tags = obj.tags.map(String);
    } else if (typeof obj.tags === 'string') {
      store.tags = obj.tags
        .split(/[;,|]/)
        .map((s: string) => s.trim())
        .filter(Boolean);
    }
  }

  if (obj.dayId !== undefined) {
    store.dayId = String(obj.dayId);
  }

  if (obj.openHours !== undefined) {
    const oh = obj.openHours as PlainObj;
    const parsed: StoreOpenHours = {};
    for (const [day, windows] of Object.entries(oh)) {
      const key = day.toLowerCase();
      if (!['mon', 'tue', 'wed', 'thu', 'fri', 'sat', 'sun'].includes(key)) {
        throw new Error(`Invalid day in openHours for store ${id}: ${day}`);
      }
      if (!Array.isArray(windows)) {
        throw new Error(`openHours for store ${id} day ${day} must be an array`);
      }
      parsed[key as Weekday] = windows.map((w) => {
        if (!Array.isArray(w) || w.length !== 2) {
          throw new Error(
            `openHours window for store ${id} day ${day} must be [open, close]`,
          );
          }
        const [open, close] = w.map(String);
        if (!/^\d{2}:\d{2}$/.test(open) || !/^\d{2}:\d{2}$/.test(close)) {
          throw new Error(
            `Invalid time format in openHours for store ${id} day ${day}`,
          );
        }
        if (hhmmToMin(close) <= hhmmToMin(open)) {
          throw new Error(
            `close must be later than open for store ${id} day ${day}`,
          );
        }
        return [open, close] as [string, string];
      });
    }
    store.openHours = parsed;
  }

  return store;
}

function parseDay(obj: PlainObj): DayConfig {
  if (!obj || typeof obj.dayId !== 'string') {
    throw new Error('Day must have dayId');
  }
  const win = obj.window as PlainObj;
  const day: DayConfig = {
    dayId: obj.dayId,
    start: parseAnchor(obj.start as PlainObj),
    end: parseAnchor(obj.end as PlainObj),
    window: { start: String(win.start), end: String(win.end) },
  };

  if (obj.mph !== undefined) {
    day.mph = Number(obj.mph);
  }
  if (obj.defaultDwellMin !== undefined) {
    day.defaultDwellMin = Number(obj.defaultDwellMin);
  }
  if (obj.mustVisitIds) {
    if (!Array.isArray(obj.mustVisitIds)) {
      throw new Error('mustVisitIds must be an array');
    }
    day.mustVisitIds = obj.mustVisitIds.map(String);
  }
  if (obj.locks) {
    day.locks = obj.locks as LockSpec[];
  }

  if (obj.maxDriveTime !== undefined) {
    day.maxDriveTime = Number(obj.maxDriveTime);
  }
  if (obj.maxStops !== undefined) {
    day.maxStops = Number(obj.maxStops);
  }
  if (obj.breakWindow) {
    const bw = obj.breakWindow as PlainObj;
    day.breakWindow = { start: String(bw.start), end: String(bw.end) };
  }
  if (obj.robustnessFactor !== undefined) {
    day.robustnessFactor = Number(obj.robustnessFactor);
  }
  if (obj.riskThresholdMin !== undefined) {
    day.riskThresholdMin = Number(obj.riskThresholdMin);
  }

  if (obj.dayOfWeek !== undefined) {
    const map: Record<string, Weekday> = {
      mon: 'mon',
      monday: 'mon',
      tue: 'tue',
      tuesday: 'tue',
      wed: 'wed',
      wednesday: 'wed',
      thu: 'thu',
      thursday: 'thu',
      fri: 'fri',
      friday: 'fri',
      sat: 'sat',
      saturday: 'sat',
      sun: 'sun',
      sunday: 'sun',
    };
    const key = String(obj.dayOfWeek).toLowerCase();
    const code = map[key];
    if (!code) {
      throw new Error(`Invalid dayOfWeek: ${obj.dayOfWeek}`);
    }
    day.dayOfWeek = code;
  }

  return day;
}

function parseTripConfig(obj?: PlainObj): TripConfig {
  const cfg: TripConfig = {};
  if (!obj) return cfg;
  if (obj.mph !== undefined) cfg.mph = Number(obj.mph);
  if (obj.defaultDwellMin !== undefined)
    cfg.defaultDwellMin = Number(obj.defaultDwellMin);
  if (obj.seed !== undefined) cfg.seed = Number(obj.seed);
  if (obj.snapDuplicateToleranceMeters !== undefined)
    cfg.snapDuplicateToleranceMeters = Number(
      obj.snapDuplicateToleranceMeters,
    );
  if (obj.robustnessFactor !== undefined)
    cfg.robustnessFactor = Number(obj.robustnessFactor);
  if (obj.riskThresholdMin !== undefined)
    cfg.riskThresholdMin = Number(obj.riskThresholdMin);
  return cfg;
}

function parseCsvStores(csv: string): PlainObj[] {
  const rows: PlainObj[] = [];
  const lines = csv
    .trim()
    .split(/\r?\n/)
    .filter((l) => l.trim().length > 0);
  if (lines.length === 0) return rows;
  const headers = lines[0].split(',').map((h) => h.trim());
  for (let i = 1; i < lines.length; i++) {
    const values = lines[i].split(',');
    const obj: PlainObj = {};
    headers.forEach((h, idx) => {
      const v = values[idx];
      if (v !== undefined) obj[h] = v.trim();
    });
    rows.push(obj);
  }
  return rows;
}

/**
 * Parse trip JSON (and optional CSV list of stores) into typed structures with validation.
 */
export function parseTrip(json: PlainObj, storesCsv?: string): TripInput {
  if (typeof json !== 'object' || json === null) {
    throw new Error('Trip JSON must be an object');
  }

  const config = parseTripConfig(json.config as PlainObj);
  const days = Array.isArray(json.days)
    ? json.days.map((d) => parseDay(d as PlainObj))
    : [];

  let storeObjs: PlainObj[] = Array.isArray(json.stores)
    ? [...(json.stores as PlainObj[])]
    : [];
  if (storesCsv) {
    storeObjs = storeObjs.concat(parseCsvStores(storesCsv));
  }
  let stores = storeObjs.map(parseStore);

  // Validate unique IDs across anchors and stores
  const seen = new Set<string>();
  const checkUnique = (id: string) => {
    if (seen.has(id)) {
      throw new Error(`Duplicate id: ${id}`);
    }
    seen.add(id);
  };
  for (const day of days) {
    checkUnique(day.start.id);
    checkUnique(day.end.id);
  }
  for (const store of stores) {
    checkUnique(store.id);
  }

  const tol = config.snapDuplicateToleranceMeters;
  if (tol && tol > 0) {
    const toleranceMiles = tol / 1609.344;
    const deduped: Store[] = [];
    for (const s of stores) {
      const match = deduped.find(
        (d) => haversineMiles(d.coord, s.coord) <= toleranceMiles,
      );
      if (match) {
        if (match.id !== s.id) {
          console.warn(
            `Dropping store ${s.id} at ${s.coord} as duplicate of ${match.id}`,
          );
        }
        continue; // skip duplicate
      }
      deduped.push(s);
    }
    stores = deduped;
  }

  const storeById = new Map<string, Store>();
  for (const s of stores) storeById.set(s.id, s);

  // Validate must-visit IDs exist
  for (const day of days) {
    if (!day.mustVisitIds) continue;
    for (const id of day.mustVisitIds) {
      const store = storeById.get(id);
      if (!store) {
        throw new Error(`Must-visit id not found: ${id}`);
      }
      if (store.dayId && store.dayId !== day.dayId) {
        throw new Error(
          `Must-visit id ${id} not available on day ${day.dayId}`,
        );
      }
    }
  }

  return { config, days, stores };
}

