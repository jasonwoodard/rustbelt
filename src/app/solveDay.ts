import { emitItinerary, EmitResult } from '../io/emit';
import { solveCommon, augmentErrorWithReasons } from './solveCommon';
import type { LockSpec, DayPlan } from '../types';
import type { ProgressFn } from '../heuristics';

export interface SolveDayOptions {
  tripPath: string;
  dayId: string;
  mph?: number;
  defaultDwellMin?: number;
  seed?: number;
  verbose?: boolean;
  locks?: LockSpec[];
  progress?: ProgressFn;
  lambda?: number;
  robustnessFactor?: number;
  riskThresholdMin?: number;
}

export interface SolveDayResult extends EmitResult {
  metrics: DayPlan['metrics'];
}

export function solveDay(opts: SolveDayOptions): SolveDayResult {
  try {
    const dayPlan = solveCommon({
      tripPath: opts.tripPath,
      dayId: opts.dayId,
      mph: opts.mph,
      defaultDwellMin: opts.defaultDwellMin,
      seed: opts.seed,
      verbose: opts.verbose,
      locks: opts.locks,
      progress: opts.progress,
      lambda: opts.lambda,
      robustnessFactor: opts.robustnessFactor,
      riskThresholdMin: opts.riskThresholdMin,
    });
    const emit = emitItinerary([dayPlan]);
    return { ...emit, metrics: dayPlan.metrics };
  } catch (err) {
    throw augmentErrorWithReasons(err);
  }
}

