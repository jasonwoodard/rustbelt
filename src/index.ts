import { Command } from 'commander';
import { writeFileSync, mkdirSync, readFileSync } from 'node:fs';
import { dirname } from 'node:path';
import { solveDay } from './app/solveDay';
import { reoptimizeDay } from './app/reoptimizeDay';
import { emitKml } from './io/emitKml';
import { emitCsv } from './io/emitCsv';
import { emitHtml } from './io/emitHtml';
import { parseTrip } from './io/parse';
import type { DayPlan, ID } from './types';
import type { ProgressFn } from './heuristics';

function buildProgressLogger(verbose: boolean): ProgressFn {
  return (phase, order, metrics) => {
    const detail = verbose ? ` order=${order.join(',')}` : '';
    console.log(
      `progress ${phase}: stops=${order.length} slack=${metrics.slackMin.toFixed(
        1,
      )} drive=${metrics.totalDriveMin.toFixed(1)} eta=${metrics.hotelETAmin.toFixed(1)}${detail}`,
    );
  };
}

export const program = new Command();

program
  .name('rustbelt')
  .description('CLI for ...')
  .version('0.1.0')
  .showHelpAfterError();

program
  .command('solve-day',  { isDefault: true })
  .requiredOption('--trip <file>', 'Path to trip JSON file')
  .requiredOption('--day <id>', 'Day id to solve')
  .option('--mph <mph>', 'Average speed in mph', parseFloat)
  .option(
    '--default-dwell <min>',
    'Default dwell minutes',
    parseFloat,
  )
  .option('--seed <seed>', 'Random seed', parseFloat)
  .option('--lambda <lambda>', 'Score weighting (0=count,1=score)', parseFloat)
  .option('--verbose', 'Print heuristic steps')
  .option('--progress', 'Print heuristic progress')
  .option('--now <HH:mm>', 'Reoptimize from this time')
  .option('--at <lat,lon>', 'Current location')
  .option('--done <ids>', 'Comma-separated list of completed store IDs')
  .option('--out <file>', 'Write itinerary JSON to this path (overwrite)')
  .option('--kml [file]', 'Write KML to this path (or stdout)')
  .option('--html [file]', 'Write HTML itinerary to this path (or stdout)')
  .option('--csv <file>', 'Write store stops CSV to this path')
  .option(
    '--robustness <factor>',
    'Multiply travel times by this factor',
    parseFloat,
  )
  .option(
    '--risk-threshold <min>',
    'Slack threshold minutes for on-time risk',
    parseFloat,
  )
  .action((opts) => {
    let result;
    if (opts.now && opts.at) {
      const [lat, lon] = opts.at.split(',').map(Number);
      const completedIds = opts.done
        ? String(opts.done)
            .split(',')
            .map((s) => s.trim())
            .filter(Boolean)
        : undefined;
      result = reoptimizeDay(
        opts.now,
        [lat, lon],
        {
          tripPath: opts.trip,
          dayId: opts.day,
          mph: opts.mph,
          defaultDwellMin: opts.defaultDwell,
          seed: opts.seed,
          lambda: opts.lambda,
          verbose: opts.verbose,
          completedIds,
          progress: opts.progress
            ? buildProgressLogger(Boolean(opts.verbose))
            : undefined,
          robustnessFactor: opts.robustness,
          riskThresholdMin: opts['risk-threshold'],
        },
      );
    } else {
      result = solveDay({
        tripPath: opts.trip,
        dayId: opts.day,
        mph: opts.mph,
        defaultDwellMin: opts.defaultDwell,
        seed: opts.seed,
        lambda: opts.lambda,
        verbose: opts.verbose,
        progress: opts.progress
          ? buildProgressLogger(Boolean(opts.verbose))
          : undefined,
        robustnessFactor: opts.robustness,
        riskThresholdMin: opts['risk-threshold'],
      });
    }

    const data = JSON.parse(result.json) as { days: DayPlan[] };

    if (opts.out) {
      mkdirSync(dirname(opts.out), { recursive: true });
      writeFileSync(opts.out, result.json, 'utf8');
      console.log(`Wrote ${opts.out}`);
    }

    if (opts.csv) {
      const runTs = new Date().toISOString();
      const rawTrip = readFileSync(opts.trip, 'utf8');
      const trip = parseTrip(JSON.parse(rawTrip));
      const mustVisitByDay: Record<string, ReadonlySet<ID>> = {};
      for (const d of trip.days) {
        if (d.mustVisitIds) {
          mustVisitByDay[d.dayId] = new Set(d.mustVisitIds);
        }
      }
      const storeMustVisitIds = new Set(
        trip.stores
          .filter((s) => (s.tags ?? []).some((t) => /must[-_]?visit/i.test(t)))
          .map((s) => s.id),
      );
      const csv = emitCsv(data.days, runTs, opts.seed ?? trip.config.seed, {
        mustVisitByDay,
        storeMustVisitIds,
      });
      mkdirSync(dirname(opts.csv), { recursive: true });
      writeFileSync(opts.csv, csv, 'utf8');
      console.log(`Wrote ${opts.csv}`);
    }

    if (opts.kml !== undefined) {
      const kml = emitKml(data.days);
      if (typeof opts.kml === 'string') {
        mkdirSync(dirname(opts.kml), { recursive: true });
        writeFileSync(opts.kml, kml, 'utf8');
        console.log(`Wrote ${opts.kml}`);
      } else {
        console.log(kml);
      }
    }

    if (opts.html !== undefined) {
      const html = emitHtml(data.days);
      if (typeof opts.html === 'string') {
        mkdirSync(dirname(opts.html), { recursive: true });
        writeFileSync(opts.html, html, 'utf8');
        console.log(`Wrote ${opts.html}`);
      } else {
        console.log(html);
      }
    }

    console.log(result.json);
  });
  
export function run(argv: readonly string[] = process.argv): Command {
  program.parse(argv as string[]);
  return program;
}

import { fileURLToPath } from 'node:url';

if (process.argv[1] && fileURLToPath(import.meta.url) === process.argv[1]) {
  run();
}
