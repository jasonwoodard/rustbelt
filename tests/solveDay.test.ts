import { describe, it, expect } from 'vitest';
import { solveDay } from '../src/app/solveDay';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';
import { readFileSync } from 'node:fs';
import { parseTrip } from '../src/io/parse';
import { computeTimeline } from '../src/schedule';
import type { Store } from '../src/types';
import { hhmmToMin } from '../src/time';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

describe('solveDay', () => {
  it('produces a day plan for a simple trip', () => {
    const tripPath = join(__dirname, '../fixtures/simple-trip.json');
    const result = solveDay({ tripPath, dayId: 'D1' });
    const data = JSON.parse(result.json);
    expect(typeof data.runTimestamp).toBe('string');
    const ids = data.days[0].stops.map((s: { id: string }) => s.id);
    expect(ids).toEqual(['S', 'A', 'B', 'C', 'E']);
    expect(data.days[0].metrics.storesVisited).toBe(3);
    expect(data.days[0].metrics.totalScore).toBe(0);
  });

  it('includes runId and runNote when provided', () => {
    const tripPath = join(__dirname, '../fixtures/run-id-note-trip.json');
    const raw = readFileSync(tripPath, 'utf8');
    const trip = parseTrip(JSON.parse(raw));
    expect(trip.config.runId).toBe('RID');
    expect(trip.config.runNote).toBe('RN');
    const result = solveDay({ tripPath, dayId: 'D1' });
    const data = JSON.parse(result.json);
    expect(data.runId).toBe('RID');
    expect(data.runNote).toBe('RN');
  });

  it('produces stable itinerary output (FR-31)', () => {
    const tripPath = join(__dirname, '../fixtures/simple-trip.json');
    const result = solveDay({ tripPath, dayId: 'D1' });
    const data = JSON.parse(result.json);
    delete data.runTimestamp;
    expect(data).toMatchSnapshot();
  });

  it('respects store scores when lambda=1', () => {
    const tripPath = join(__dirname, '../fixtures/scored-trip.json');
    const resCount = solveDay({ tripPath, dayId: 'D1' });
    const dataCount = JSON.parse(resCount.json);
    const storesCount = dataCount.days[0].stops
      .filter((s: { type: string }) => s.type === 'store')
      .map((s: { id: string }) => s.id);
    expect(storesCount).toEqual(['A']);

    const resScore = solveDay({ tripPath, dayId: 'D1', lambda: 1 });
    const dataScore = JSON.parse(resScore.json);
    const storesScore = dataScore.days[0].stops
      .filter((s: { type: string }) => s.type === 'store')
      .map((s: { id: string }) => s.id);
    expect(storesScore).toEqual(['B']);
    expect(dataScore.days[0].metrics.totalScore).toBe(100);
  });

  it('throws if must visits exceed day window', () => {
    const tripPath = join(
      __dirname,
      '../fixtures/infeasible-must-visit.json',
    );
    const raw = readFileSync(tripPath, 'utf8');
    const trip = parseTrip(JSON.parse(raw));
    const day = trip.days[0];
    const stores: Record<string, Store> = {};
    for (const s of trip.stores) {
      stores[s.id] = s;
    }
    const ctx = {
      start: day.start,
      end: day.end,
      window: day.window,
      mph: day.mph ?? trip.config.mph ?? 30,
      defaultDwellMin:
        day.defaultDwellMin ?? trip.config.defaultDwellMin ?? 0,
      stores,
    };
    const { hotelETAmin } = computeTimeline(day.mustVisitIds!, ctx);
    const endMin = hhmmToMin(day.window.end);
    const deficit = Math.round(hotelETAmin - endMin);
    try {
      solveDay({ tripPath, dayId: 'D1' });
      throw new Error('expected solveDay to throw');
    } catch (err) {
      const e = err as Error;
      expect(e.message).toContain(
        `must visits exceed day window by ${deficit} min`,
      );
      expect(e.message).toContain('Must-visit chain exceeds window');
      const match = e.message.match(/suggestions: (.*)$/);
      expect(match).not.toBeNull();
      const suggestions = JSON.parse(match![1]);
      const types = suggestions.map((s: { type: string }) => s.type);
      expect(types).toContain('extendEnd');
      expect(types).toContain('dropMustVisit');
      expect(
        suggestions.every(
          (s: { reason?: string }) => typeof s.reason === 'string',
        ),
      ).toBe(true);
    }
  });

  it('throws when start→end drive equals window', () => {
    const tripPath = join(__dirname, '../fixtures/drive-equals-window.json');
    expect(() => solveDay({ tripPath, dayId: 'D1' })).toThrow(
      'start to end drive time 60 min >= window 60 min',
    );
  });

  it('plans when start→end drive is shorter than window', () => {
    const tripPath = join(__dirname, '../fixtures/drive-within-window.json');
    const result = solveDay({ tripPath, dayId: 'D1' });
    const data = JSON.parse(result.json);
    const ids = data.days[0].stops.map((s: { id: string }) => s.id);
    expect(ids).toEqual(['S', 'E']);
  });
});
