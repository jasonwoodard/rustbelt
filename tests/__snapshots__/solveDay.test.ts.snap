// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`solveDay > produces stable itinerary output (FR-31) 1`] = `
{
  "days": [
    {
      "dayId": "D1",
      "metrics": {
        "slackMin": 748.0590557204848,
        "storesVisited": 3,
        "totalDriveMin": 690.9409442795152,
        "totalDwellMin": 0,
      },
      "stops": [
        {
          "arrive": "00:00",
          "depart": "00:00",
          "id": "S",
          "name": "start",
          "type": "start",
        },
        {
          "arrive": "02:18",
          "depart": "02:18",
          "dwellMin": 0,
          "id": "A",
          "legIn": {
            "distanceMi": 138.18818885590304,
            "driveMin": 138.18818885590304,
            "fromId": "S",
            "toId": "A",
          },
          "name": "A",
          "type": "store",
        },
        {
          "arrive": "05:45",
          "depart": "05:45",
          "dwellMin": 0,
          "id": "B",
          "legIn": {
            "distanceMi": 207.28228328385455,
            "driveMin": 207.28228328385455,
            "fromId": "A",
            "toId": "B",
          },
          "name": "B",
          "type": "store",
        },
        {
          "arrive": "09:13",
          "depart": "09:13",
          "dwellMin": 0,
          "id": "C",
          "legIn": {
            "distanceMi": 207.28228328385458,
            "driveMin": 207.28228328385458,
            "fromId": "B",
            "toId": "C",
          },
          "name": "C",
          "type": "store",
        },
        {
          "arrive": "11:31",
          "depart": "11:31",
          "id": "E",
          "legIn": {
            "distanceMi": 138.188188855903,
            "driveMin": 138.188188855903,
            "fromId": "C",
            "toId": "E",
          },
          "name": "end",
          "type": "end",
        },
      ],
    },
  ],
}
`;
